keps/349-avoid-coscheduling/README.md

```diff
@@ -0,0 +1,350 @@
+workload.
+
+My use can be supported by adding `kueue.x-k8s.io/wait-for-pod-scheduling`
+annotation to the jobs I run.
+
+#### Story 2
+
+As an administrator of the batch processing platform I want to be able to
+configure kueue quotas for on-demand resources a little higher than what is
+available in the quiet period. However, I don't want to risk two workloads
+deadlock in case of issues with provisioning nodes to satisfy the configured
+quota.
+
+My use can be supported by adding `kueue.x-k8s.io/wait-for-pod-scheduling`
+annotation to all jobs. I could also designate a single local queue
+`wait-for-pod-scheduling-queue` to be responsible for such workloads and only
+add the annotation based on the value of the `kueue.x-k8s.io/queue-name`
+annotation. This way we could allow for coscheduling of workloads if the local
+queue is not used.
+
+### Notes/Contraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+If the workload fails to schedule its pods it could block admission of other
+workloads indefinitely.
+
+In order to mitigate this risk we terminate workloads which don't have enough
+pods ready after the timeout.
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+
+### Ready workload condition
+
+First, we introduce a new workload condition, called `Ready`, to denote workloads
+which have their pods scheduled, more precisely, which satisfy the condition:
+`job.status.ready + job.status.succeeded >= job.spec.parallelism`.
+The condition is added by the kueue's Job controller when it reconciles a job on
+a Job status update. Note that, verifying the condition does not require extra
+API calls to the job or individual pods.
+
+### Workload API
+
+We extend the Workload API in order to specify the strategy
+
+```golang
+// WorkloadSpec defines the desired state of Workload
+type WorkloadSpec struct {
+  ...
+  // waitForPodsScheduled is true indicates that the workload blocks admission
+  // of all other workloads until it is in the Ready condition. Further, it
+  // delays its corresponding job start until all other admitted workloads
```

Comment:
```suggestion
  // delays admission of this workload until all other admitted workloads
```

Reply:
Actually, I meant what is written. We want to delay the job start if there are any other admitted workloads which are still not ready.